File: android/src/main/java/com/riguz/encryptions/cipher/Argon2.java
package com.riguz.encryptions.cipher;

public final class Argon2 {
    static {
        System.loadLibrary("argon2-binding");
    }

    private final int iterations;
    private final int memory;
    private final int parallelism;
    private final int hashLength;

    public Argon2(int iterations, int memory, int parallelism, int hashLength) {
        this.iterations = iterations;
        this.memory = memory;
        this.parallelism = parallelism;
        this.hashLength = hashLength;
    }

    public byte[] argon2i(final byte[] password, final byte[] salt) {
        return argon2iInternal(iterations, memory, parallelism, password, salt, hashLength);
    }

    public byte[] argon2d(final byte[] password, final byte[] salt) {
        return argon2dInternal(iterations, memory, parallelism, password, salt, hashLength);
    }

    private native byte[] argon2iInternal(int iterations, int memory, int parallelism, final byte[] password, final byte[] salt, int hashLength);

    private native byte[] argon2dInternal(int iterations, int memory, int parallelism, final byte[] password, final byte[] salt, int hashLength);
}

File: ios/Classes/Argon2.swift
import Foundation

struct Argon2 {
    private let iterations: UInt32;
    private let memory: UInt32;
    private let parallelism: UInt32;
    private let hashLength: UInt32;
    
    init(iterations: UInt32,
         memory: UInt32,
         parallelism: UInt32,
         hashLength: UInt32){
        self.iterations = iterations;
        self.memory = memory;
        self.parallelism = parallelism;
        self.hashLength = hashLength;
    }
    
    func argon2i(password: Data, salt: Data)-> Data {
        var outputBytes  = [UInt8](repeating: 0, count: Int(hashLength));
        
        password.withUnsafeBytes { passwordBytes in
            salt.withUnsafeBytes {
                saltBytes in
                argon2i_hash_raw(iterations, memory, parallelism, passwordBytes, password.count, saltBytes, salt.count, &outputBytes, Int(hashLength));
            }
        }
        
        return Data(bytes: UnsafePointer<UInt8>(outputBytes), count: Int(hashLength));
    }
    
    func argon2d(password: Data, salt: Data)-> Data {
        var outputBytes  = [UInt8](repeating: 0, count: Int(hashLength));
        
        password.withUnsafeBytes { passwordBytes in
            salt.withUnsafeBytes {
                saltBytes in
                argon2d_hash_raw(iterations, memory, parallelism, passwordBytes, password.count, saltBytes, salt.count, &outputBytes, Int(hashLength));
            }
        }
        
        return Data(bytes: UnsafePointer<UInt8>(outputBytes), count: Int(hashLength));
    }
    
    func argon2id(password: Data, salt: Data)-> Data {
        var outputBytes  = [UInt8](repeating: 0, count: Int(hashLength));
        
        password.withUnsafeBytes { passwordBytes in
            salt.withUnsafeBytes {
                saltBytes in
                argon2id_hash_raw(iterations, memory, parallelism, passwordBytes, password.count, saltBytes, salt.count, &outputBytes, Int(hashLength));
            }
        }
        
        return Data(bytes: UnsafePointer<UInt8>(outputBytes), count: Int(hashLength));
    }
}
